### ISSUE TITLE
Take back control of git-hooks (aka The Great Migration)

### ISSUE THREAD

**Issue by bkcsoft** [go-gitea/gitea#343]
- Gitea version (or commit ref): c8f300b
- Git version: 2.10.2

This is in no way related to https://github.com/gogits/gogs/issues/1623 ...

Basically I wanna move most of the git-related callbacks into git-hooks, this would however break all existing support for custom git-hooks, but it will make it easier to implement features such as https://github.com/go-gitea/gitea/issues/302 https://github.com/go-gitea/gitea/issues/32 https://github.com/go-gitea/gitea/issues/8 and would fix https://github.com/go-gitea/gitea/issues/218 without actually having to fix it (since the bug is that Gitea have to duplicate the input-data to shove one down gits throat and the other into what is essentially `.git/hooks/update` 😒 )

thoughts?

**Comment by tboerger** [comment]
I would also take the control of the hooks but that doesn't directly solve the mentioned issues :)

**Comment by strk** [comment]
Breaking existing deploys won't help adoption of Gitea, unless
automatic adaption of hooks is provided.

**Comment by tboerger** [comment]
This is intended for 2.0.0 where breaking changes will be totally fine.

**Comment by tboerger** [comment]
As this will introduce real breaking changes I have created and assigned the 2.x.x milestone.

**Comment by bkcsoft** [comment]
@strk possible migrations involve calling user provided hooks at the end of giteas hooks, which should be fairly easy to make a migration for 🙂 

@tboerger it says "make it easier to implement" not "fix" 😛 the only thing it will automagically fix is #218 which only exists as a bug since we're _not_ using git-hooks for what git-hooks was created for 😆

**Comment by adymitruk** [comment]
what's the workaround for getting a custom update hook? Add this to the repos' update hooks? 
```
if [ -e hooks/update.custom ]; then bash hooks/update.custom "$@"; fi || exit
# the regular gitea line
```

**Comment by lunny** [comment]
I don't think this should be a v2.x.x issue. It should be fixed ASAP.

**Comment by bkcsoft** [comment]
@lunny potential breakage, loads of migrations and backwards compatibility. In my book it belongs in a major-release :slightly_smiling_face:

**Comment by tboerger** [comment]
Somebody can build a POC than we can see if it should work without breaking installations or not :)

**Comment by bkcsoft** [comment]
I'll have a look at cherry-picking and correcting a few things from https://github.com/gogits/gogs/commit/039dc333670a7ece5b34a611b794d072b589ec65 during the weekend 🙂

**Comment by lunny** [comment]
resolved by #1006
